### PR TITLE
Add TextSet and update TextClassification examples

### PR DESCRIPTION
**Pull request by hkvision**
The second PR separated from #601 with modifications.
- Add TextSet.
- Update TextClassification examples using TextSet and Transformers.
- Fit, predict and evaluate on TextSet.
- Add some missing doc for Python TextFeature.

Some remarks:
- Directly reading using sc is slower in the preprocessing step (collecting word frequencies) than reading without sc and transforming to DistributedTextSet afterwards.
- Will add predictor in a separate PR.

### CODE REVIEW

**Comment by hkvision** [comment]
This PR is almost finished. Please have a review, thanks! @jason-dai @dding3

**Comment by hkvision** [comment]
jenkins passed http://10.239.47.211:18888/view/ZOO-PR/job/ZOO-PR-Validation/1126/

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+  override def isLocal: Boolean = true
+
+  override def isDistributed: Boolean = false
+
+  override def toLocal(): LocalTextSet = {
+    this
+  }
+
+  override def toDistributed(sc: SparkContext): DistributedTextSet = {
+    new DistributedTextSet(sc.parallelize(array))
```
Comment (dding3) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: why do we need support convert a localTextSet to DistributedTextSet?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+  }
+
+  override def toDistributed(sc: SparkContext, partitionNum: Int): DistributedTextSet = {
```
Comment (dding3) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: Kind of repeated with the previous api?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    rdd = transformer(rdd)
+    this
+  }
+
+  override def isLocal: Boolean = false
+
+  override def isDistributed: Boolean = true
+
+  override def toLocal(): LocalTextSet = {
+    new LocalTextSet(rdd.collect())
```
Comment (dding3) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: why do we need support convert this?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 4]
```diff
@@ -109,6 +109,7 @@ abstract class KerasNet[T: ClassTag](implicit ev: TensorNumeric[T])
       loss: Criterion[T],
       metrics: List[ValidationMethod[T]] = null)(implicit ev: TensorNumeric[T]): Unit = {
     LoggerFilter.redirectSparkInfoLogs()
+    Logger4j.getLogger("com.intel.analytics.bigdl.optim").setLevel(Level4j.INFO)
```
Comment (dding3) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: Can we remove Logger4j.getLogger and LoggerFilter.redirectSparkInforLogs? I think we can leave it set by user?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 5]
```diff
@@ -143,6 +144,7 @@ abstract class KerasNet[T: ClassTag](implicit ev: TensorNumeric[T])
       loss: (Variable[T], Variable[T]) => Variable[T],
       metrics: List[ValidationMethod[T]])(implicit ev: TensorNumeric[T]): Unit = {
     LoggerFilter.redirectSparkInfoLogs()
+    Logger4j.getLogger("com.intel.analytics.bigdl.optim").setLevel(Level4j.INFO)
```
Comment (dding3) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: Same as above.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 4 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: Thanks. I have reverted the change of getLogger.
But if user forget to call `LoggerFilter.redirectSparkInfoLogs()` in their application, during the training process, there will be tons of info printed out and it is really hard to find the loss/accuracy info that the user really want to see. I think it will be better to call this by default? @dding3

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala [hunk 6]
```diff
@@ -142,38 +79,15 @@ object TextClassification {
...
-          featureTensor = Tensor(input, Array(sequenceLength)),
-          label = label)
-      }
-
-      val Array(trainingRDD, valRDD) = sampleRDD.randomSplit(
-        Array(trainingSplit, 1 - trainingSplit))
+      val textSet = TextSet.read(param.baseDir + "/20news-18828/")
+        .toDistributed(sc, param.partitionNum)
+      println("Processing text dataset...")
+      val transformed = textSet.tokenize().normalize().shapeSequence(param.sequenceLength)
```
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala]: We can do this... One extra thing we need to do is to generate the wordIndex map, which should be obtained from the TextSet while the transformer doesn't contain such information.
We can probably write the code like the following:
```
val tokenized = textSet -> Tokenizer() -> Normalizer() -> SequenceShaper()
val map = tokenized.generateWordIndexMap()
val transformed = tokenized -> WordIndexer(map) -> TextFeatureToSample()
```

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 2 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: Just to provide the option of `partitionNum`?

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala]: It looks good to me.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 2 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: I think this api is good enough, with a default partitionNum value, it can cover previous api?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+   *
+   * @param weights Array of Double indicating the split portions.
+   */
+  def randomSplit(weights: Array[Double]): Array[TextSet]
+
+  /**
+   * Do tokenization on original text.
+   * See Tokenizer for more details.
+   */
+  def tokenize(): TextSet = {
```
[the remark this answers is not in the document]
Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: I think we should add these commonly used text/image specific transformations to TextSet and ImageSet.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 2 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: Yeah, I removed the previous api and gives partitionNum default =4 in this api, is it ok?

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala [hunk 8]
```diff
@@ -142,38 +81,15 @@ object TextClassification {
...
-      val sampleRDD = indexedRdd.map {case (input: Array[Float], label: Float) =>
-        Sample(
-          featureTensor = Tensor(input, Array(sequenceLength)),
-          label = label)
-      }
-
-      val Array(trainingRDD, valRDD) = sampleRDD.randomSplit(
-        Array(trainingSplit, 1 - trainingSplit))
+      val textSet = TextSet.read(param.baseDir + "/20news-18828/")
+        .toDistributed(sc, param.partitionNum)
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala]: Using `TextSet.read(dir, sc, partitionNum)` can also work but slower than reading as a LocalTextSet and then converting to a distributed one.

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala]: Users can definitely write in this way. Alternatively, users are recommended to use `.tokenize()...`, which will be more convenient.

**Comment by hkvision** [comment]
jenkins passed http://10.239.47.211:18888/view/ZOO-PR/job/ZOO-PR-Validation/1149/

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 2 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: Sounds good to me

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: OK. Will add them to ImageSet soon.

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala]: Ok

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+   *             All texts will be given a label according to the subdirectory where it is located.
+   *             Labels start from 0.
+   * @param sc An instance of SparkContext if any. Default is null.
+   * @param minPartitions A suggestion value of the minimal partition number.
+   *                      Integer. Default is 1. Only need to specify this when sc is not null.
+   * @return TextSet.
+   */
+  def read(path: String, sc: SparkContext = null, minPartitions: Int = 1): TextSet = {
+    val textSet = if (sc != null) {
+      val fs = FileSystem.get(new Configuration())
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: need test on hdfs

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 4 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: I have raised an issue and will remove `LoggerFilter.redirectSparkInforLogs` in another PR.

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala [hunk 10]
```diff
@@ -142,38 +81,15 @@ object TextClassification {
...
-        .map {case (tokens, label) => (shaping(tokens, sequenceLength), label)}
-      val sampleRDD = indexedRdd.map {case (input: Array[Float], label: Float) =>
-        Sample(
-          featureTensor = Tensor(input, Array(sequenceLength)),
-          label = label)
-      }
-
-      val Array(trainingRDD, valRDD) = sampleRDD.randomSplit(
-        Array(trainingSplit, 1 - trainingSplit))
+      val textSet = TextSet.read(param.baseDir + "/20news-18828/")
```
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/textclassification/TextClassification.scala]: I suppose we may not use a new dataset. But to be safe and make users more flexible, I will make `20` an options and separate arguments of data path and embedding path.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+      }
+      rdd.persist()
+      val wordIndex = TextSet.wordIndexFromFrequencies(frequencies)
+      val wordIndexBC = rdd.sparkContext.broadcast(wordIndex)
+      setWordIndex(wordIndexBC.value)
+      wordIndexBC.value
+    } catch {
+      case s: SparkException =>
+        TextSet.logger.warn("TextSet hasn't been tokenized yet, please call tokenize() first")
+        null
```
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: I replaced by printing the stack trace.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/python/PythonTextFeature.scala [hunk 12]
```diff
@@ -118,4 +122,184 @@ class PythonTextFeature[T: ClassTag](implicit ev: TensorNumeric[T]) extends Pyth
...
+      TextSet.read(path, null, minPartitions)
+    }
+    else {
+      TextSet.read(path, sc.sc, minPartitions)
+    }
+  }
+
+  def textSetGetWordIndex(textSet: TextSet): JMap[String, Int] = {
+    val res = textSet.getWordIndex
+    if (res == null) {
```
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/python/PythonTextFeature.scala]: This is Python syntax? In Scala it should be `if (res == null) null else res.asJava`. I just removed the curly braces.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+
+  override def generateWordIndexMap(
+    removeTopN: Int = 0, maxWordsNum: Int = -1): Map[String, Int] = {
+    try {
+      var frequencies = rdd.flatMap(_.getTokens).filter(_ != "##")  // "##" is the padElement.
+        .map(word => (word, 1)).reduceByKey(_ + _).sortBy(- _._2).collect().drop(removeTopN)
+      if (maxWordsNum > 0) {
+        frequencies = frequencies.take(maxWordsNum)
+      }
+      rdd.persist()
```
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: It is a bit tricky to do unpersist, I will move this first as in the original version it isn't considered. Can investigate this later if it is necessary.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 9 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: @hkvision Manually tested and works for HDFS

**Comment by hkvision** [comment]
@dding3 Manually tested for reading news20 from HDFS.

**Comment by hkvision** [comment]
@dding3 All the comments are addressed :) Please have a final look at this PR if you have time and I myself will have a look as well. I'm going to merge it soon. :)

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 3 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: Ok

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: I think we don't need catch the exception here, just throw the exception?

**Comment by hkvision** [comment]
jenkins passed http://10.239.47.211:18888/view/ZOO-PR/job/ZOO-PR-Validation/1185/
Thanks so much for your review and help! @dding3
